`flaps`: retry on certain failure modes, such as 504s

alichay opened the pull request:
### Change Summary

What and Why: All flaps operations will retry if the error is known to be transient. This should improve reliability - especially in situations like `fly deploy` where one hiccup can (currently) stop an entire deploy.

How: Adds a wrapper in `flapsutil` over `*flaps.Client` that implements retrying on certain failure modes, like 504. `NewClientWithOptions` now returns the generic `FlapsClient` so that we can return the wrapper type instead of a raw `*flaps.Client`.

---

### Documentation

- [x] Fresh Produce
- [ ] In superfly/docs, or asked for help from docs team
- [ ] n/a

Commit:
`machine`: stub out `Retry` and `RetryRet`

These functions allow retrying machine operations like machine creation, though there's no logic yet for determining which errors are *actually* transient, so behavior is still the same

Commit:
`machine`: local backoff state instead of global

backoff modifies the Backoff state object while using it, so having one globally makes me uncomfortable

Commit:
`machine`: add 504 error to `Retry`

Commit:
`machine`: cleanup how error retryability is determined

Commit:
`machine`: `Retry` now supports timeouts

Commit:
`machine`: `Retry`: strongly typed flaps error handling

Commit:
`machine`: use `Retry` in most `internal/machine` functions

except for ephemeral machines - i'm cooking up a different strategy to do this so this is more of a check-in than anything

Commit:
Revert "`machine`: use `Retry` in most `internal/machine` functions"

This reverts commit 49e38ba27b7ab1e1e001269ce04ac047dc31bd5a.

Commit:
Revert "`machine`: `Retry` now supports timeouts"

This reverts commit fb76c768d29fc96ffa55964ca009bf184bc54b3f.

Commit:
`flaps`: exclusively use the `FlapsClient` interface over `flaps.Client`

Commit:
`flaps`: wrap the Flaps client with a retryer

Since Flaps has been abstracted to an Interface in flyctl, we can add a retry layer to it by just wrapping the existing client with *few* changes made elsewhere

This is just a drop-in reliability improvement

rugwirobaker commented:
Retrying POST/PATCH my yield unexpected results cause subsequent requests might to go to a different flaps and flyd among other issues. It's basically not Idempotent. https://flyio.discourse.team/t/flaps-what-status-codes-can-we-retry/5060/2

alichay commented:
> Retrying POST/PATCH my yield unexpected results cause subsequent requests might to go to a different flaps and flyd among other issues. It's basically not Idempotent. https://flyio.discourse.team/t/flaps-what-status-codes-can-we-retry/5060/2

ah. that's a huge bummer. I'm going to try to see if I can get this working for GET requests then, and maybe afterwards come up with a different strategy for `fly deploy` machine creation.

I wonder if some operations, like `SetMetadata`, could be retried anyway? it's setting a named value, so even if it were called multiple times there shouldn't be any side effects...

Commit:
`retryClient`: restrict retry to GET and ops that should be idempotent

Going to have to follow up to see whether all of these are safe to retry, but I feel like they *should* be?
